README
README file for Perl extension XML::SAX::ExpatXS
================================================

1. Introduction
2. Dependencies
3. Callbacks
4. Encoding


1.Introduction
------------------------------

This module is a direct XS implementation of Perl SAX parser 
using Expat. XML::SAX::Expat is implemented as a layer over 
XML::Parser.

The first version of this module has been created by Matt 
Sergeant reusing sources of XML::Parser. The current maintainer 
is Petr Cimprich <petr AT gingerall DOT cz>.

The wrapper is considered stable. Feedback of any kind is 
appreciated.


2. Dependencies
------------------------------

This module requires Expat XML parser. XML::SAX::ExpatXS 1.10 
or higher requires Expat 1.95.4 or higher (because 
of XML_SetSkippedEntityHandler). For older Expat versions, it 
is recommended to use XML::SAX::ExpatXS 1.09. 

Required Perl modules are XML::Namespace Support and XML::SAX.


3. Callbacks
------------------------------

These PerlSAX callbacks are supported and tested:

 start_document()
 end_document()
 start_element()
 end_element()
 characters()
 processing_instruction()
 start_prefix_mapping()
 end_prefix_mapping()
 set_document_locator()
 fatal_error()
 comment()
 start_dtd()
 end_dtd()
 start_cdata()
 end_cdata()
 element_decl
 attribute_decl
 notation_decl()
 unparsed_entity_decl()
 external_entity_decl()
 internal_entity_decl()
 start_entity()
 end_entity ()
 resolve_entity()
 skipped_entity()

These methods are never called by XML::SAX::ExpatXS:

 warning()
 error()
 ignorable_whitespace()


This one is deprecated but if works with XML::SAX::ExpatXS:

 xml_decl()


4. Encoding
------------------------------

These charsets and encodings are supported:

UTF-8        (1)
UTF-16	     (1)
US-ASCII     (1)
ISO-8859-1   (1)
ISO-8859-2   (2)
ISO-8859-3   (2)
ISO-8859-4   (2)
ISO-8859-5   (2)
ISO-8859-7   (2)
ISO-8859-8   (2)
ISO-8859-9   (2)
WINDOWS-1250 (2)
WINDOWS-1252 (2)
BIG5	     (2)
EUC-KR	     (2)
EUC-JP	     (2,3)
Shift JIS.   (2,3)

(1) Expat built-in
(2) external handler
(3) see lib/XML/SAX/ExpatXS/Encodings/Japanese_Encodings.msg

Other encodings can be added with XML::Encoding, see 
lib/XML/SAX/ExpatXS/Encodings/README for more info.


Changes
Revision history for Perl extension XML::SAX::ExpatXS
=====================================================

1.32 Apr 12, 2011
	- NoExpand option test added
	- type warnings fixed (gcc 4.4)

1.31 July 3, 2007
    - Security fix: namespace separator in XML_ParserCreate_MM
      changed from '}' to -1, related to SF.net bug 1742315 
      [suggested by Bjoern Hoehrmann]

1.30 July 2, 2007
    - speed optimizations - main callback functions (start_element,
      end_element, characters, comment) resolved in advance using 
      XML::SAX::ExpatXS::Preload module
    - feature 'http://xmlns.perl.org/sax/recstring' controls whether
      a recognized string is available as {RecognizedString} among 
      {ParseOptions} (not by default as this is rather expensive).
    - added {ExpatVersion} property

1.20 April 7, 2006
    - speed optimizations - data structures reused: start_element() 
      hash reused by end_element() call when free, characters() hash 
      recycled when free
    - better error message when parse_uri fails

1.10 November 10, 2005
    - fixed start_entity() called for an external DTD subset if 
      parameter entity parsing is in effect (ParseParamEnt option
      setting XML_PARAM_ENTITY_PARSING_UNLESS_STANDALONE in Expat)
      [using patches by Axel Eckenberger]
    - resolve_entity() can return source with {String} now
    - feature 'http://xml.org/sax/features/external-general-entities'
      controls processing of external general entities
    - feature 'http://xml.org/sax/features/external-parameter-entities'
      controls processing of external parameter entities (the same as 
      ParseParamEnt option)
    - skipped_entity() implemented
      [based on patch and tests by Axel Eckenberger]
    - fixed segfault for unparsed entities without PublicId
      [using patches by Axel Eckenberger]
    - fixed leaks in entity callbacks

1.09 October 14, 2005
    - {Source}{Encoding} is used (with higher priority 
      than encoding specified in the XML declaration)
    - fixed segfaults for undefined {Source} properties
    - Expat localization improved in Makefile.PL
    - localized setting of $/ in _get_external_entity

1.08 April 22, 2005
    - feature 'http://xmlns.perl.org/sax/xmlns-uris'
      puts xmlns:* attributes into 
      the 'http://www.w3.org/2000/xmlns/' namespace
      and xmlns attributes into no namespace.
      The 'http://xml.org/sax/features/xmlns-uris'
      applies only if the new one is off.
      The default behavior is the same as for other 
      Perl SAX parsers now.
    - fixed repeated parsing with Perl 5.6

1.07 March 16, 2005
    - feature  'http://xml.org/sax/features/xmlns-uris'
      puts xmlns and xmlns:* attributes into no or
      the 'http://www.w3.org/2000/xmlns/' namespace.
      Default is no namespace (feature not set) which
      is a behavior different from previous versions!
    - feature 'http://xmlns.perl.org/sax/version-2.1'
      is set (and can't be turned off)  
    - minor fix in memory management
      [reported by Rob Bloodgood] 

1.06 March 7, 2005
    - more changes in memory management; 
      parse_* run in loops without leaks
      [reported by Rob Bloodgood]

1.05 February 18, 2005
    - fixed segfault after deleting a key of Features
      [patch by Pavel Hlavnicka]

1.04 December 15, 2004
    - fixed the bug of SystemId/PublicId of external DTD 
      subset provided instead of identifiers of the current 
      entity in Locator and Exception
      [reported by Bjoern Hoehrmann]
    - recognized string made available
      as $p->{ParseOptions}{RecognizedString}
    - ParserOptions renamed to ParseOptions to match 
      XML::SAX::Base
    - minor fixes for make test
    - test suite extended

1.03 November 23, 2004
    - minor fixes for make test

1.02 November 22, 2004
    - document locator redesigned to fit
      Perl SAX 2.1 specs
    - feature 'http://xmlns.perl.org/sax/locator' can turn
      the document locator off
    - test suite extended

1.01 September 24, 2004
    - memory management check completed, 
      no more leaks detected

1.00 July 16, 2004
    - relevant memory leaks fixed
    - some more XS code cleanup 

0.99 June 7, 2004
    - element_decl (Name, Model) completed
    - attribute_decl (eName, aName, Type, Mode, Value) completed
    - ns declaration attributes are treated as common attributes
      when the 'http://xmlns.perl.org/sax/ns-attributes' feature
      is set to 1 (default)
    - fixed the bug of "0" translated into ""
      [patch by Sam Tregar]
    - fixed a reference counting bug in startElement
      [reported by Sam Tregar]
    - test suite extended

0.98 April 9, 2004
    - added support for more encodings. In addition to
      built-in Expat encodings (UTF-8, UTF-16, ISO-8859-1, 
      US-ASCII) we now support ISO-8859-2,3,4,5,7,8,9,
      WIN-1250,1252, BIG5, EUC-KR, EUC-JP and Shift JIS.
      (adapted from XML::Parser)
    - cleaning the code to satisfy gcc2 and ANSI C
      [suggested by Sergey Skvortsov]
    - minor bug fixes (locator, test suite)

0.97 March 29, 2004
    - external parsed entities parsed
    - external_entity_decl (Name, PublicId, SystemId) called 
      for external entity declarations
    - internal_entity_decl (Name, Value) called for external 
      entity declarations
    - start_entity (Name), end_entity (Name) called for 
      external entity references
    - resolve_entity callback implemented
    - get_feature(), set_feature(), get_features() API functions 
      implemented 
    - feature 'http://xmlns.perl.org/sax/join-character-data' controls 
      character data fragmentation, 0/1 (1 being default)
    - performance optimizations
    - fixed the xmlns="" segmentation fault
    - test suite extended

0.96 February 19, 2004
    - exceptions (Message, Exception, LineNumber, 
      ColumnNumber, PublicId, SystemId) sent to ErrorHandler
    - document locator (LineNumber, ColumnNumber, PublicId, 

Makefile.PL
# This Makefile.PL contains portions borrowed 
# from XML::Parser and XML::SAX

use ExtUtils::MakeMaker;
use Config;

$expat_libpath = '';
$expat_incpath = '';

my @replacement_args;

foreach (@ARGV) {
  if (/^EXPAT(LIB|INC)PATH=(.+)/) {
    if ($1 eq 'LIB') {
      $expat_libpath = $2;
    }
    else {
      $expat_incpath = $2;
    }
  }
  else {
    push(@replacement_args, $_);
  }
}

@ARGV = @replacement_args;

if (not $expat_libpath and $] >= 5.006001) {
  require ExtUtils::Liblist;		# Buggy before this
  ($expat_libpath) = ExtUtils::Liblist->ext('-lexpat');
}

unless ($expat_libpath) {
  # Test for existence of libexpat
  my $found = 0;
  foreach (split(/\s+/, $Config{libpth})) {
    if (-f "$_/libexpat." . $Config{so}) {
      $found = 1;
      last;
    }
  }

  unless ($found) {
      die <<'Expat_Not_Installed;';

Expat must be installed prior to building XML::SAX::ExpatXS and I can't 
find it in the standard library directories. You can download expat from:

http://sourceforge.net/projects/expat/

If expat is installed, but in a non-standard directory, then use the
following options to Makefile.PL:

    EXPATLIBPATH=...  To set the directory in which to find libexpat

    EXPATINCPATH=...  To set the directory in which to find expat.h

For example:

    perl Makefile.PL EXPATLIBPATH=/home/me/lib EXPATINCPATH=/home/me/include

Note that if you build against a shareable library in a non-standard location
you may (on some platforms) also have to set your LD_LIBRARY_PATH environment
variable at run time for perl to find the library.

Expat_Not_Installed;
  }
}

my $libs = "-lexpat";
$libs = "-L$expat_libpath $libs"
  if $expat_libpath;

@extras = ();

push(@extras, INC => "-I$expat_incpath")
  if $expat_incpath;

push(@extras,
     CAPI => 'TRUE')
    if ($PERL_VERSION >= 5.005 and $OSNAME eq 'MSWin32'
	and $Config{archname} =~ /-object\b/i);

push(@extras,
     ABSTRACT_FROM => 'lib/XML/SAX/ExpatXS.pm',
     AUTHOR => 'Matt Sergeant <matt AT sergeant DOT org>')
    if ($ExtUtils::MakeMaker::Version >= 5.4301);


WriteMakefile(
    NAME => 'XML::SAX::ExpatXS',
    VERSION_FROM => 'lib/XML/SAX/ExpatXS.pm',
    PREREQ_PM => {
		    'XML::SAX' => 0.13,
		   },
    LIBS => $libs,
    @extras
);

sub MY::install {
    package MY;
    my $script = shift->SUPER::install(@_);
    if (ExtUtils::MakeMaker::prompt("Do you want to alter ParserDetails.ini?", "Y") =~ /^y/i) {
        $script =~ s/install :: (.*)$/install :: $1 install_expat_xs/m;
        $script .= <<"INSTALL";

install_expat_xs :
\t\@\$(PERL) -MXML::SAX -e "XML::SAX->add_parser(q(XML::SAX::ExpatXS))->save_parsers()"

INSTALL

    }

    return $script;
}


lib/XML/SAX/ExpatXS.pm
package XML::SAX::ExpatXS;
use strict;
use vars qw($VERSION @ISA);

use XML::SAX::ExpatXS::Encoding;
use XML::SAX::ExpatXS::Preload;
use XML::SAX::Base;
use DynaLoader ();
use Carp;
use IO::File;

$VERSION = '1.32';
@ISA = qw(DynaLoader XML::SAX::Base XML::SAX::ExpatXS::Preload);

XML::SAX::ExpatXS->bootstrap($VERSION);

my @features = (
	['http://xml.org/sax/features/namespaces', 1],
	['http://xml.org/sax/features/external-general-entities', 1],
	['http://xml.org/sax/features/external-parameter-entities', 0],
    ['http://xml.org/sax/features/xmlns-uris', 0],
    ['http://xmlns.perl.org/sax/xmlns-uris', 1],
    ['http://xmlns.perl.org/sax/version-2.1', 1],
	['http://xmlns.perl.org/sax/join-character-data', 1],
	['http://xmlns.perl.org/sax/ns-attributes', 1],
	['http://xmlns.perl.org/sax/locator', 1],
	['http://xmlns.perl.org/sax/recstring', 0]
			 );
my @supported_features = map($_->[0], @features);


#------------------------------------------------------------
# API methods
#------------------------------------------------------------

sub new {
    my $proto = shift;
    my $options = ($#_ == 0) ? shift : { @_ };

    foreach (@features) {
	$options->{Features}->{$_->[0]} = $_->[1];
    }

    $options->{ExpatVersion} = ExpatVersion();

    return $proto->SUPER::new($options);
}

sub get_feature {
    my ($self, $feat) = @_;
      if (exists $self->{Features}->{$feat}) {
	  return $self->{Features}->{$feat};
      }
      else {
          return $self->SUPER::get_feature($feat);
      }
  }

sub set_feature {
    my ($self, $feat, $val) = @_;
      if (exists $self->{Features}->{$feat}) {
	  return $self->{Features}->{$feat} = $val;
      }
      else {
          return $self->SUPER::set_feature($feat, $val);
      }
  }

sub get_features {
    my $self = shift;
    return %{$self->{Features}};
}

sub supported_features {
    my $self = shift;

    return @supported_features;
}

#------------------------------------------------------------
# internal methods
#------------------------------------------------------------

sub _parse_characterstream {
    my ($self, $fh) = @_;
    $self->{ParseOptions}->{ParseFunc} = \&ParseStream;
    $self->{ParseOptions}->{ParseFuncParam} = $fh;
    $self->_parse;
}

sub _parse_bytestream {
    my ($self, $fh) = @_;
    $self->{ParseOptions}->{ParseFunc} = \&ParseStream;
    $self->{ParseOptions}->{ParseFuncParam} = $fh;
    $self->_parse;
}

sub _parse_string {
    my ($self, $str) = @_;
    $self->{ParseOptions}->{ParseFunc} = \&ParseString;
    $self->{ParseOptions}->{ParseFuncParam} = $str;
    $self->_parse;
}

sub _parse_systemid {
    my ($self, $uri) = @_;
    my $fh = IO::File->new($uri) or croak "ExpatXS: Can't open $uri ($!)";
    $self->{ParseOptions}->{ParseFunc} = \&ParseStream;
    $self->{ParseOptions}->{ParseFuncParam} = $fh;
    $self->_parse;
}

sub _parse {
    my $self = shift;

    my $args = bless $self->{ParseOptions}, ref($self);
    delete $args->{ParseOptions};

    # copy handlers over
    $args->{Handler} = $self->{Handler};
    $args->{DocumentHandler} = $self->{DocumentHandler};
    $args->{ContentHandler} = $self->{ContentHandler};
    $args->{DTDHandler} = $self->{DTDHandler};
    $args->{LexicalHandler} = $self->{LexicalHandler};
    $args->{DeclHandler} = $self->{DeclHandler};
    $args->{ErrorHandler} = $self->{ErrorHandler};
    $args->{EntityResolver} = $self->{EntityResolver};

    $args->{_State_} = 0;
    $args->{Context} = [];
    $args->{ErrorMessage} ||= '';
    $args->{Namespace_Stack} = [[ xml => 'http://www.w3.org/XML/1998/namespace' ]];
    $args->{Parser} = ParserCreate($args, 
				   $args->{Source}{Encoding} 
				   || $args->{ProtocolEncoding}, 
				   1);
    $args->{Locator} = GetLocator($args->{Parser}, 
				  $args->{Source}{PublicId} || '',
				  $args->{Source}{SystemId} || '',
				  $args->{Source}{Encoding} || '',
				 );
    $args->{RecognizedString} = GetRecognizedString($args->{Parser});
    $args->{ExternEnt} = GetExternEnt($args->{Parser});

    $args->{Methods} = {};
    $args->get_start_element();
    $args->get_end_element();
    $args->get_characters();
    $args->get_comment();

    # the most common handlers are available as refs
    SetCallbacks($args->{Parser}, 
		 $args->{Methods}->{start_element},
		 $args->{Methods}->{end_element},
		 $args->{Methods}->{characters},
		 $args->{Methods}->{comment},
  		);

    $args->set_document_locator($args->{Locator});
    $args->start_document({});
   
    my $result;
    $result = $args->{ParseFunc}->($args->{Parser}, $args->{ParseFuncParam});

    ParserFree($args->{Parser});

    my $rv = $args->end_document({});   # end_document is still called on error

    croak($args->{ErrorMessage}) unless $result;
    return $rv;
}

sub _get_external_entity {
    my ($self, $base, $sysid, $pubid) = @_;

    # resolving with the base URI
    if ($base and $sysid and $sysid !~ /^[a-zA-Z]+[a-zA-Z\d\+\-\.]*:/) {
	$base =~ s/[^\/]+$//;
	$sysid = $base . $sysid;
    }

    # user defined resolution
    my $src = $self->resolve_entity({PublicId => $pubid, 
				     SystemId => $sysid});
    my $fh;
    my $result;
    my $string;
    if (ref($src) eq 'CODE') {
	$fh = IO::File->new($sysid)
	  or croak("Can't open external entity: $sysid\n");

    } elsif (ref($src) eq 'HASH') {
	if (defined $src->{CharacterStream}) {
	    $fh = $src->{CharacterStream};

	} elsif (defined $src->{ByteStream}) {
	    $fh = $src->{ByteStream};

	} elsif (defined $src->{String}) {
	    $result = $src->{String};
	    $string = 1;

	} else {
	    $fh = IO::File->new($src->{SystemId})
	      or croak("Can't open external entity: $src->{SystemId}\n");
	}

    } else {
	croak ("Invalid object returned by EntityResolver: $src\n");
    }

    unless ($string) {
	local $/;
	undef $/;
	$result = <$fh>;
	close($fh);
    }
    return $result;
}

sub _get_handler_methods {
    my $self = shift;


}

1;
__END__

=head1 NAME

XML::SAX::ExpatXS - Perl SAX 2 XS extension to Expat parser

=head1 SYNOPSIS

 use XML::SAX::ExpatXS;

 $handler = MyHandler->new();
 $parser = XML::SAX::ExpatXS->new( Handler => $handler );
 $parser->parse_uri($uri);
  #or
 $parser->parse_string($xml);

=head1 DESCRIPTION

XML::SAX::ExpatXS is a direct XS extension to Expat XML parser. It implements
Perl SAX 2.1 interface. See http://perl-xml.sourceforge.net/perl-sax/ for
Perl SAX API description. Any deviations from the Perl SAX 2.1 specification 
are considered as bugs.

=head2 Features

The parser behavior can be changed by setting features.

 $parser->set_feature(FEATURE, VALUE);

XML::SAX::ExpatXS provides these adjustable features:

=over

=item C<http://xmlns.perl.org/sax/join-character-data>

Consequent character data are joined (1, default) or not (0).

=item C<http://xmlns.perl.org/sax/ns-attributes>

Namespace attributes are reported as common attributes (1, default) or not (0).

=item C<http://xmlns.perl.org/sax/xmlns-uris>

When set on, xmlns and xmlns:* attributes are put into namespaces in a Perl SAX
traditional way; xmlns attributes are in no namespace while xmlns:* attributes
are in the C<http://www.w3.org/2000/xmlns/> namespace. This feature is set to 1
by default.

=item C<http://xml.org/sax/features/xmlns-uris>

This feature applies if and only if the C<http://xmlns.perl.org/sax/xmlns-uris>
feature is off. Then, xmlns and xmlns:* attributes are both put into no namespace 
(0, default) or into C<http://www.w3.org/2000/xmlns/> namespace (1).

=item C<http://xmlns.perl.org/sax/locator>

The document locator is updated (1, default) for ContentHadler events or not (0).

=item C<http://xmlns.perl.org/sax/recstring>

A recognized string (the text string currently processed by this XML parser) 
is either maintained as $parser->{ParseOptions}{RecognizedString} (1) or not 
(0, default).

=item C<http://xml.org/sax/features/external-general-entities>

Controls whether this parser processes external general entities (1, default)
or not (0).

=item C<http://xml.org/sax/features/external-parameter-entities>

Controls whether this parser processes external parameter entities including 
an external DTD subset (1) or not (0, default).

=back

=head2 Constructor Options

Apart from features, the behavior of this parser can also be changed with
options to the constructor.

=over

=item ParseParamEnt

 ParseParamEnt => 1

This option meaning is exactly the same as 
the C<http://xml.org/sax/features/external-parameter-entities> feature. 
The option is supported only because of the compatibility with older versions
of this module. Turned off by default.

=item NoExpand

 NoExpand => 1

No internal entities are expanded if this option is turned on.
Turned off by default.

=back

=head2 Read-only Properties

=over

=item ExpatVersion

This property returns a version of linked Expat library, for example 
expat_1.95.7.

=back

=head1 AUTHORS

 Petr Cimprich <petr AT gingerall DOT org> (maintainer)
 Matt Sergeant <matt AT sergeant DOT org>

=cut
